Add OnTileConverted hook

Tyfyter opened the pull request:
### What is the new feature?
Added ``ModTile.OnTileConverted`` and ``GlobalTile.OnTileConverted``. which run after a tile is converted by ``WorldGen.Convert``.

### Why should this be part of tModLoader?
Modders may wish to track tiles the player converts, make a biome that reacts to being converted, or otherwise do something which requires running some code when a tile is converted.

### Sample usage for the new feature
```cs
public class ExampleOnTileConverted : GlobalTile {
	public override void OnTileConverted(int i, int j, int fromType, int toType, int conversionType) {
		if (TileID.Sets.Corrupt[toType]) {
			Main.NewText("Hello, cruel world!");
		} else if (TileID.Sets.Corrupt[fromType]) {
			Main.NewText("Goodbye, cruel world!");
		}
	}
}
```

Tyfyter committed:
Added OnTileConverted hook

Chicken-Bones commented:
Assuming it works, lgtm, add one for walls too

Tyfyter commented:
Tested by adding
```public override void OnTileConverted(int i, int j, int fromType, int toType, int conversionType) {
	switch ((TileID.Sets.Corrupt[toType] || TileID.Sets.Crimson[toType]).ToInt() - (TileID.Sets.Corrupt[fromType] || TileID.Sets.Crimson[fromType]).ToInt()) {
		case 1:
			Main.NewText($"Hello, cruel world! {TileID.Search.GetName(fromType)} --> {TileID.Search.GetName(toType)} (converted by {conversionType})");
			break;
		case 0:
			Main.NewText($"I guess you're here, cruel world! {TileID.Search.GetName(fromType)} --> {TileID.Search.GetName(toType)} (converted by {conversionType})");
			break;
		case -1:
			Main.NewText($"Goodbye, cruel world! {TileID.Search.GetName(fromType)} --> {TileID.Search.GetName(toType)} (converted by {conversionType})");
			break;
	}
}``` to `CorruptFossilTile`, fixed issues which arose in testing.

Tyfyter committed:
OnWallConverted & fixes

Tyfyter committed:
fixed recursion counter not decrementing on error

Tyfyter committed:
Fixed trigger on direct calls to Tile/WallLoader.Convert

Tyfyter committed:
Trigger on hardUpdateWorld

Tyfyter committed:
made ConversionRecursion readonly struct

JavidPack committed:
Merge branch '1.4.4' into pr/4778

JavidPack committed:
Update WorldGen.cs.patch

JavidPack committed:
Merge branch '1.4.4' into pr/4778

JavidPack committed:
Documentation, formatting, typos.

JavidPack committed:
Update WorldGen.cs.patch

JavidPack committed:
Update WorldGen.cs.patch

Tyfyter committed:
Update WorldGen.cs.patch

Tyfyter committed:
Support GERunner

Chicken-Bones committed:
Merge branch '1.4.4' into 1.4.4-OnTileConverted-hook

Chicken-Bones commented on patches/tModLoader/Terraria/ModLoader/TileLoader.cs:
```diff
@@ -875,6 +875,7 @@ public static bool Convert(int i, int j, int conversionType)
 
 		if (tile.TileType == type && TryGetConversionFallback(type, conversionType, out var fallback)) {
 			tile.TileType = (ushort)fallback;
+			using var recursionCounter = new WorldGen.ConversionRecursion();
```
Why? It seems to me that anything calling `Tile/WallLoader.Convert` should have already established a `ConversionRecursion` so that they can call `OnTileConverted` at the end of the vanilla code

Tyfyter replied:
Nothing but ``WorldGen.Convert`` needs to check for recursion before calling the hook because they can't have been called by ``Tile/WallLoader.Convert``.

Chicken-Bones replied:
But if something else in a modded `TileLoader.Convert` hook calls `WorldGen.Convert` the `OnTileConverted` hook will be called twice, because `WorldGen.Convert` will not know that there is already a conversion in progress

Tyfyter replied:
I'm not sure I understand the situation you're saying it could happen in fully because the only interpretation I can think of in which that could happen seems highly impractical and *should* have OnTileConverted be called, but this one line could be moved to the beginning of the method to stop that

Chicken-Bones commented on patches/tModLoader/Terraria/Projectile.cs.patch:
```diff
@@ -1555,21 +1555,32 @@
++						int tileType = Main.tile[num96, num97].type;
++						int wall = Main.tile[num96, num97].wall;
  						if (type == 10) {
 +							if (!TileLoader.Convert(num96, num97, BiomeConversionID.PurificationPowder))
```
You should create and use a conversion recursion context, incase some mod calls `WorldGen.Convert` as part of their hook

Tyfyter replied:
Fixed in https://github.com/tModLoader/tModLoader/pull/4778/commits/4cbdeb9eca2388a0792d1809d1df6293679bcf4e

Chicken-Bones commented on patches/tModLoader/Terraria/WorldGen.cs.patch:
```diff
@@ -3013,6 +3054,7 @@
 +
 +					if (preHookType != Main.tile[num8, num9].type) {
 +						flag = true;
++						TileLoader.OnTileConverted(num8, num9, preHookType, Main.tile[num8, num9].type, BiomeConversionID.Chlorophyte);
```
This doesn't support all the vanilla conversions. You could merge this into the `else if` chain below with something like:
```cs
if (preHookType != Main.tile[num8, num9].type) {
    flag = true;
}
else if (!convertTile)
{}
else
if (Main.tile[num8, num9].type == 661 || Main.tile[num8, ...
```

Chicken-Bones commented on patches/tModLoader/Terraria/WorldGen.cs.patch:
```diff
@@ -1996,19 +2006,23 @@
 +						if (type >= TileLoader.TileCount || wall >= WallLoader.WallCount)
- 							continue;
++							goto SkipTilePurityConversion;
++							//continue;
```
These vanilla bounds checks are silly and could just be commented out entirely.
```cs
						/*
						if (type > TileID.Count || wall > WallID.Count)
							continue;
						*/
```

Chicken-Bones commented on patches/tModLoader/Terraria/WorldGen.cs.patch:
```diff
@@ -3234,10 +3276,34 @@
++					ushort oldWallType = Main.tile[m, n].wall;
++
 +					bool convertWall = WallLoader.Convert(m, n, good ? BiomeConversionID.Hallow : crimson ? BiomeConversionID.Crimson : BiomeConversionID.Corruption);
 +					bool convertTile = TileLoader.Convert(m, n, good ? BiomeConversionID.Hallow : crimson ? BiomeConversionID.Crimson : BiomeConversionID.Corruption);
```
See above about using a `RecursionContext`, but is there a reason why these can't use `WorldGen.Convert`? Do they differ in some way?

Tyfyter replied:
Yes, [the hardmode V converts a few more tile types than normal conversion](https://terraria.wiki.gg/wiki/Hardmode_conversion#Converted_tiles), notably silt and — in worlds with over 200,000 of them as a measure to allow it to work in "not the bees" worlds — [Hive](https://terraria.wiki.gg/wiki/Hive) blocks, we would need to reimplement 9 conversions in order to preserve these while switching to ``WorldGen.Convert``.

Tyfyter committed:
Never call OnTile/WallConverted twice

Calling ``WorldGen.Convert`` from within a ``Tile/WallLoader.Convert`` hook can arguably never have intuitive behavior, chose to just never call the hooks twice

Tyfyter committed:
Support vanilla conversions from Chlorophyte

Chicken-Bones committed:
Patch cleanup
